Bug 1691241 [csc] Prevent getting stuck in current state when cache is invalid for csc

kevinrizza opened the pull request:
Problem:
When the service restarts, we lose all cached knowledge about the
catalogsourceconfig, including the old targetNamespace. This is an
issue, because if the targetNamespace changed we want to remove the
catalogsource in the old targetNamespace and replace it with a new one.
Now that we are always assuming the cache is invalid if it is empty
(which is a valid assumption), we still don't have a way of cleaning up
old targetNamespace if it changed. Right now though, we will return an
error and not attempt to reconcile the CSC.

Not only that, but this error would also cause the catalogsourceconfig
to get into a stuck state. When the error was returned, we moved back
into the current state to retry. If the state was `success` or `failure`
the `catalogsourceconfig` would simply loop in that phase until the
status was dropped or the syncer set the status to initial.

Solution:
If the cache is invalid we cannot assume that the targetNamespace is
actually invalid. For now, we should just print a debug log and return
and continue to reconcile the csc. Once we have a better method
of retaining references to owned objects, we can handle this edge case
better.

Additionally, update the `deleteObjects` function to no longer return
an error. We do not want to cycle forever attempting to clean up the
`catalogsource` if we are unable to, since that would cause our
`catalogsourceconfig` to be stuck in a bad state looping on the state
until the syncer restarted the reconcilliation loop or through manual
intervention.

This issue was uncovered by the original fix for
https://bugzilla.redhat.com/show_bug.cgi?id=1691241, and this commit
is addition to that fix.

aravindhp commented on pkg/catalogsourceconfig/update.go:
```diff
@@ -62,10 +61,17 @@ func (r *updateReconciler) Reconcile(ctx context.Context, in *v1alpha1.CatalogSo
 	cachedCSCSpec, found := r.cache.Get(in)
-	// This should never happen as it is because the cached Spec has changed
+	// This generally won't happen as it is because the cached Spec has changed
 	// that we are in the update reconciler.
```
Nit: `when we are in`

kevinrizza replied:
Updated.

aravindhp commented on pkg/catalogsourceconfig/update.go:
```diff
@@ -62,10 +61,17 @@ func (r *updateReconciler) Reconcile(ctx context.Context, in *v1alpha1.CatalogSo
+	// we should be using something like a direct reference to the old object
+	// instead of an in memory cache, since when the service restarts we will be
+	// unable to find the old object if the targetNamespace changed while we were
+	// down. For now, we will just ignore it.
```
Like we discussed offline, this doesn't seem to be the issue @emmajiafan is talking about in the BZ.

kevinrizza replied:
This is actually not the case, we are seeing this error because the cache error forces the csc into a bad state that won't be resolved unless the status is dropped. By preventing this bad state, we won't have to wait for the syncer to help us.

tkashem commented on pkg/catalogsourceconfig/update.go:
```diff
@@ -62,10 +61,17 @@ func (r *updateReconciler) Reconcile(ctx context.Context, in *v1alpha1.CatalogSo
+	// instead of an in memory cache, since when the service restarts we will be
+	// unable to find the old object if the targetNamespace changed while we were
+	// down. For now, we will just ignore it.
 	if !found {
```
Question: I am trying to understand what real world scenario is causing this if block to be executed. We detected that the target namespace has changed ( `r.targetChanged` == true ) but the spec is not in cache.

kevinrizza replied:
@tkashem because the cache invalidator logic changed. Previously we were cheating by saying that if we didn't find the `csc` in the cache that we should assume that everything is fine and ignore the cache. To be able to force a resync I updated that logic to assume that if the `csc` wasn't in the cache that the cache is not valid (which is an accurate assumption).

The problem is, we genuinely don't know if the targetNamespace was changed if we lose track of the cache. So we should say it's invalid and attempt to cleanup, but we don't have a way of finding the old one.

tkashem replied:
Yeah, if the cache is missing then we can't answer the question whether the target namespace has changed. We will need to persist the namespaced name of `catsrc` in the future in `annotations` or something.

kevinrizza commented:
@aravindhp @tkashem addressed comments, please take another look.

tkashem commented:
/lgtm

kevinrizza commented:
@aravindhp @tkashem I removed the unused error returned from the `updateReconciler`'s `deleteObjects()` function, which is technically what caused the endless reconcile loop described in the BZ. I also updated the comments in the source and the git commit to reflect that change.

kevinrizza commented:
/test e2e-aws

kevinrizza commented:
/retest

kevinrizza commented:
/retest

kevinrizza commented:
/retest

kevinrizza commented:
/retest

kevinrizza commented:
@tkashem @aravindhp can someone also give this PR an lgtm so that can merge once the tests are complete?

aravindhp commented:
I already gave `/approve`. @tkashem please review and give an lgtm.

tkashem commented:
/lgtm
